### PR TITLE
[MNT]: create build-requirements.txt and update dev-requirements.txt 

### PR DESCRIPTION
**Pull request by abhi-jha**
## PR summary
Adding explicit docs to make sure the project installs in editable mode

Ref : https://matrix.to/#/!BXmyZMTnRjWJldDRLV:gitter.im/$iAbowa1uqTBT5Kvdqhvn_0hj9w1g0s0E7BOduGgv57g?via=gitter.im&via=matrix.org&via=cadair.com

## PR checklist
- [ ] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [ x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

File: doc/devel/development_setup.rst
Comment (abhi-jha) [reply doc/devel/development_setup.rst:181]: Are these docs for building the docs still valid : https://matplotlib.org/devdocs/devel/document.html

I see the make commands don't work.

Comment (abhi-jha) [reply doc/devel/development_setup.rst:181]: I would like to have a local iteration before committing.

File: doc/devel/development_setup.rst
Comment (timhoffm) [reply doc/devel/development_setup.rst:182]: I don't think we should/need to mention dependencies here again. They should be set up in the two previous sections. 

The setup is a multi-step process, which we assume is done in sequence. We also would not write in this section again "make sure that you have cloned the repo".

If anything is missing concerning dependencies, it should be add to https://matplotlib.org/devdocs/devel/development_setup.html#create-a-dedicated-environment or https://matplotlib.org/devdocs/devel/development_setup.html#install-dependencies.

Comment (abhi-jha) [reply doc/devel/development_setup.rst:182]: @timhoffm Are the dev dependencies enough to setup matplotlib in dev mode? 

`pip install -r requirements/dev/dev-requirements.txt` 

If I remember from our conversation with @story645 it is not very clear on which dependencies to install and that's what we need an explicit note for.

File: doc/devel/development_setup.rst
Comment (abhi-jha) [reply doc/devel/development_setup.rst:181]:
```
(.venv) ➜  matplotlib git:(abhishek/doc-fix) ✗ make html

make: *** No rule to make target `html'.  Stop.

```

Comment (abhi-jha) [reply doc/devel/development_setup.rst:181]: Wait, looks like I need to go into the doc/ folder to execute these. Seems to have picked up.

File: doc/devel/development_setup.rst
Comment (abhi-jha) [reply doc/devel/development_setup.rst:182]: Yeah if that's the case then we need to include build deps somewhere or reference transitively. 

Can you reference how is conda doing it? I mean we are sure at this point that the pip dev dependencies don't have it.

Comment (abhi-jha) [reply doc/devel/development_setup.rst:182]: >make a new ci-requirements file with the contents of dev-requirements 

ok

>find and replace so ci knows to use it

I am confused. Why would we want to mess with the CI builds? The problem was while trying to install it locally in an editable mode.

>add the new build-requirements file to the list in dev-requirements.

Not sure I understand. We change the CI build req. Add it to dev req?

**Comment by abhi-jha** [comment]
@story645 Need some clarifications above.

**Comment by OdileVidrine** [comment]
For anyone who is having the issue "No module named setuptools_scm" after doing "pip install pybind11 meson-python cmake ninja numpy", the fix is " pip install setuptools-scm".

File: doc/devel/development_setup.rst
Comment (tacaswell) [reply doc/devel/development_setup.rst:182]: We should be clear that the reason the dependencies need to be manually pre-installed is the `--no-build-isolation` not the `-e`.

Comment (abhi-jha) [reply doc/devel/development_setup.rst]: Alright, I have made some fixes and I need one more clarification from @tacaswell 

>We should be clear that the reason the dependencies need to be manually pre-installed is the --no-build-isolation not the -e.

Can you elaborate a bit more or point me to relevant docs for why does that matter? 

Right now I have moved the dependecies to dev ones and I have pointed the CI to those requirements.txt files. 

Do you see this breaking things?

Comment (abhi-jha) [reply doc/devel/development_setup.rst]: The other way as I see being suggested is I separate `meson-python, numpy, pybind11, setuptools-scm` in a build-requirements.txt and then reference it for dev installs?

I am just curious how the current PR breaks/interferes with the setup.

Comment (abhi-jha) [reply doc/devel/development_setup.rst]: Oh wait, I think I get that we don't want all the dev dependencies to be installed in CI run.

Comment (abhi-jha) [reply doc/devel/development_setup.rst]: The following set of commands work now:

```
		python3 -m venv /tmp/clean_build
		. /tmp/clean_build/bin/activate
		pip install ".[dev]"
		pip install -ve . --no-build-isolation
```

Comment (abhi-jha) [reply doc/devel/development_setup.rst]: The CI pipeline is failing [here](https://dev.azure.com/matplotlib/matplotlib/_build/results?buildId=39506&view=logs&j=b69f81ad-e873-5389-cb38-86053e18a577&t=6b61e752-90a7-583a-d38d-8b6757c5e8cc&l=66) as they are trying to compile pybind11. This is happening because the dev dependencies are being installed in editable mode. 

Why are we installing third party dependencies in editable mode? For local development only matplotlib needs to be in editable mode.

File: requirements/dev/dev-requirements.txt
Comment (abhi-jha) [inline requirements/dev/dev-requirements.txt:1]: @tacaswell I think this is where the CI is failing as it is trying to install pybind11 in editable mode. But CI doesn't install those in editable mode. 

Does it make sense to separate this out of the -e flag?

File: doc/devel/development_setup.rst
Comment (abhi-jha) [reply doc/devel/development_setup.rst]: But there defnitely is an editable flag being applied somwhere in the CI.

File: .circleci/config.yml
Comment (abhi-jha) [reply .circleci/config.yml:106]: It is supposed to work the same way as .[dev] or .[test]. 

For example : https://github.com/matplotlib/matplotlib/blob/main/.github/workflows/cygwin.yml#L211

Comment (abhi-jha) [reply .circleci/config.yml]: Alright, made fixes. Please take a look.

File: requirements/dev/build-requirements.txt
Comment (abhi-jha) [reply requirements/dev/build-requirements.txt:1]: Oh shoot!

Comment (abhi-jha) [reply requirements/dev/build-requirements.txt:1]: Added now.

**Comment by abhi-jha** [comment]
Rebased.

**Comment by abhi-jha** [comment]
@story645 Looks good to be merged. Not sure how the review process goes. Let me know if there's something that needs to be done from my end.

**Review by timhoffm — APPROVED** [review]

File: requirements/dev/build-requirements.txt
Comment (timhoffm) [inline requirements/dev/build-requirements.txt:4]: What do we need ninja for?

Comment (abhi-jha) [reply requirements/dev/build-requirements.txt:4]: If I remember correctly I got this list in the chat on gitter while triaging the first time. We can remove it if it is unnecessary.

Comment (abhi-jha) [reply requirements/dev/build-requirements.txt]: Woah! Removing ninja has failed the CI build😮

Comment (ianthomas23) [reply requirements/dev/build-requirements.txt]: > I would have expected that meson-python declaration ninja as a dependency.

It effectively does, if you are using build isolation, but not if you aren't. See contourpy/contourpy#260 for discussion with a `meson` maintainer and follow-up in contourpy/contourpy#270.

There is no good answer to explicitly including `ninja` in the build requirements file. If you don't include it and are not using build isolation then you may well need to explicitly install it, particularly on macos and windows. But if you do include it there are situations such as on cygwin where there are no wheels available and you will end up trying to build it from source, which may not be a pleasant experience, rather than using the prebuilt cygwin package (e.g. #26543).

I have a slight preference for not including it, so that `build-requirements.txt` is consistent with the `[build-system]` section of `pyproject.toml`. This does mean that some CI runs may require it to be explicitly installed using `pip`, `sudo apt`, or whatever.

**Comment by tacaswell** [comment]
Try also removing cmake (to rely on system cmake on CI).  If the user has cmake and ninja avaialable, we should not force a reinstallation of them on top from pypi.  They should be treated/documented the same way we treat gcc?

File: requirements/dev/build-requirements.txt
Comment (abhi-jha) [reply requirements/dev/build-requirements.txt]: Do you mean that we need to put ninja dependency manually in the Ci scripts, separete from build-requirements.txt?

**Comment by abhi-jha** [comment]
> Try also removing cmake (to rely on system cmake on CI). If the user has cmake and ninja avaialable, we should not force a reinstallation of them on top from pypi. They should be treated/documented the same way we treat gcc?

Windows CI tests have failed. Is there more to it?

File: requirements/dev/build-requirements.txt
Comment (ianthomas23) [reply requirements/dev/build-requirements.txt]: We might need to, but shouldn't do it by default. We should only consider doing so in specific CI runs that fail at the build stage due to lack of `ninja`.

Comment (abhi-jha) [reply requirements/dev/build-requirements.txt]: I have removed both ninja and cmake. I see a test coverage pipeline failing. Others have passed.

**Comment by abhi-jha** [comment]
All the CI pipelines apart from the test coverage seems to have passed. How do I get the coverage to pass?

**Comment by rcomer** [comment]
@abhi-jha we are not strict about CodeCov passing https://matplotlib.org/devdocs/devel/development_workflow.html#automated-tests

This looks like it should be ready to merge as it has two approvals.  @story645 @timhoffm is there anything more to do here?

**Comment by timhoffm** [comment]
Thanks @abhi-jha!

**Pull request merged**
